L1Trigger/L1THGCal/src/be_algorithms/HGCalShowerShape.cc

```diff
@@ -207,268 +209,118 @@ float HGCalShowerShape::sigmaPhiPhiTot(const l1t::HGCalMulticluster& c3d) const
+  std::map<int, vector<float> > tc_layer_energy;
+  std::map<int, vector<float> > tc_layer_phi;
 
-        const edm::PtrVector<l1t::HGCalTriggerCell>& triggerCells = clu->constituents();
-        unsigned int ncells = triggerCells.size();
-        nTC.emplace_back(ncells);
+  const edm::PtrVector<l1t::HGCalCluster>& clustersPtrs = c3d.constituents();
 
-        for(const auto& tc : triggerCells){
+  for(const auto& clu : clustersPtrs){
+    
+    int layer = HGC_layer(clu->subdetId(),clu->layer());    
+    
+    const edm::PtrVector<l1t::HGCalTriggerCell>& triggerCells = clu->constituents();
 
-            tc_energy.emplace_back(tc->energy());
-            tc_phi.emplace_back(tc->phi());
+    for(const auto& tc : triggerCells){
 
-        }
+      tc_layer_energy[layer].emplace_back(tc->energy());
+      tc_layer_phi[layer].emplace_back(tc->phi());
 
     }
 
-    int ncl2D=layer.size();
+  }
 
-    float SigmaPhiPhiMax=0;
+  float SigmaPhiPhiMax=0;
 
-    std::vector<float> energy_layer ;
-    std::vector<float> phi_layer ;
-
-    int nLayersMax=40;//EE+FH
-    int nLayersEE=28;
-    int nLayersFH=12;
-
-    for(int ilayer=0;ilayer<nLayersMax;ilayer++){   //Loop on HGCal layers
-
-        int Layer_found=0;
-        float Layer_Spp=0;
-
-        int tc_index=0; // trigger cell index inside cl2D vector
-
-        for(int i2d=0;i2d<ncl2D;i2d++){   // Loop on cl2D inside 3DC
+  for(auto& tc_iter : tc_layer_energy){
     
-            int cl2D_layer=-999;
-
-            if(subdetID.at(i2d)==HGCEE)cl2D_layer=layer.at(i2d);
-            if(subdetID.at(i2d)==HGCHEF)cl2D_layer=layer.at(i2d)+nLayersEE;
-            if(subdetID.at(i2d)==HGCHEB)cl2D_layer=layer.at(i2d)+nLayersEE+nLayersFH;//FH
-
-            if (cl2D_layer==ilayer){
-
-                Layer_found=1; //+=1 il want to count cl2D per layer
+    vector<float> energy_layer = tc_iter.second;
+    vector<float> phi_layer= tc_layer_phi[tc_iter.first];
 
-                int ntc=nTC.at(i2d);
+    float SigmaPhiPhiLayer = sigmaPhiPhi(energy_layer,phi_layer); 
+    if(SigmaPhiPhiLayer > SigmaPhiPhiMax) SigmaPhiPhiMax = SigmaPhiPhiLayer;
 
-                for(int itc=0; itc<ntc ; itc++) {   //Loop on TC inside cl2D
+  }
 
-                    energy_layer.emplace_back(tc_energy.at(tc_index));
-                    phi_layer.emplace_back(tc_phi.at(tc_index));
-				    tc_index++;
+  return SigmaPhiPhiMax;
 
-			    }
-
-            }
-
-            else{
-
-                if (Layer_found==1) break; //Go to next ilayer
-                tc_index+=nTC.at(i2d);
-        
-            }
-
-        }
-
-
-        if(Layer_found==1) Layer_Spp=sigmaPhiPhi(energy_layer,phi_layer);    
-    
-        if(Layer_Spp>SigmaPhiPhiMax) SigmaPhiPhiMax=Layer_Spp;
 
-        energy_layer.clear();
-        phi_layer.clear();
-      
-    }    
-
-            
-    return SigmaPhiPhiMax;
-
-
-}   
+}
 
 
 
 float HGCalShowerShape::eMax(const l1t::HGCalMulticluster& c3d) const {
```

Comment:
Before the `eMax` function was computing the maximum layer energy, which is different from what is done here: the maximum cluster energy. We can have several 2D clusters in a given layer.

Reply:
Unless I'm wrong, the previous implementation was actually not leading a well-defined behavior, depending on the ordering of the cl2D when filling `std::vector<int> layer `.
If the cl2D were not ordered by layer, the following checks in the loop may have induced a break before the full energy from a layer has been computed.
`if (cl2D_layer==ilayer){`
`    Layer_found=1; //+=1 il want to count cl2D per layer`
`    Layer_Energy+=energy.at(i2d);`
`}`
`else{`
`    if (Layer_found==1) break; //Go to next ilayer`
`}`
I will update the new implementation to follow the definition you mention here.

Reply:
You're right, thanks, indeed the previous implementation was assuming ordered layers; which would give undefined results if it is not the case.